Add CHAINID specs (#168)
privacy-scaling-explorations/zkevm-specs: specs/opcode/46CHAINID.md

# CHAINID opcode

## Procedure

The `CHAINID` opcode gets the chain id where a transaction is meant executed at.

## EVM behaviour

The `CHAINID` opcode loads a 256-bit value from the chain configuration (the chain id), and pushes it to the stack.

## Circuit behaviour

1. Construct Block Context table
2. Bus-mapping lookup for stack write operation

## Constraints

1. opId == 0x46
2. State transition:
   - gc + 1 (1 stack write)
   - `stack_pointer` - 1
   - pc + 1
   - gas + 2
3. Lookups:
   - `chainid` is on the top of stack
   - `chainid` is in the block context table

## Exceptions

1. Stack overflow: stack is full, stack pointer = 0
2. out of gas: remaining gas is not enough

## Code

> Please refer to [`chainid.py`](src/zkevm_specs/evm/execution/chainid.py)

Reviewer: ```suggestion
Please refer to [`block_ctx.py`](src/zkevm_specs/evm/execution/block_ctx.py)
```

Author: Thanks for the catch!!!

# CHAINID opcode

## Procedure

The `CHAINID` opcode gets the chain id where a transaction is meant executed at.

## EVM behaviour

> The `CHAINID` opcode loads a 256-bit value from the Tx context (the chain id), and pushes it to the stack.

Reviewer 2: ```suggestion
The `CHAINID` opcode loads a 256-bit value from the chain configuration (the chain id), and pushes it to the stack.
```

Author: Ohh dammit! Since I started the specs as if the value was taken from the `tx_table`, I forgot to change this once we decided to move it to the block!!

Reviewer 3: Eventhough the implementation loads the chain_id from the block table, I think it's good to use the definition from Han (that is, "from the chain configuration"), and then explain that in our case we add this value in the block table for convenience.

## Circuit behaviour

> 1. Construct Tx Context table

Reviewer 2: ```suggestion
1. Construct Block Context table
```

2. Bus-mapping lookup for stack write operation

## Constraints

1. opId == 0x46
2. State transition:

>    - gc + 2 (1 stack write + 1 Call Context read)

Reviewer 2: ```suggestion
   - gc + 1 (1 stack write)
```

   - `stack_pointer` - 1
   - pc + 1
   - gas + 2

> 3. Lookups:
>    - `tx_id` is on current call context
>    - `chainid` is on the top of stack
>    - `chainid` is in the transaction context table

Reviewer 2: ```suggestion
3. Lookups:
   - `chainid` is on the top of stack
   - `chainid` is in the block context table
```
